File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Solution {
	public int[] solution(String[] id_list, String[] reports, int k) {
		Map<String, Set<String>> reportsByUser = new HashMap<>();
		Map<String, Integer> reportedCountByUser = new HashMap<>();
		Set<String> blockedUser = new HashSet<>();

		for (String report : reports) {
			String[] reportDetail = report.split(" ");

			boolean added = reportsByUser.computeIfAbsent(reportDetail[0], (key) -> new HashSet<>())
				.add(reportDetail[1]);

			if (added) {
				int mergedCount = reportedCountByUser.merge(reportDetail[1], 1, Integer::sum);

				if (mergedCount >= k) {
					blockedUser.add(reportDetail[1]);
				}
			}
		}

		int[] answer = new int[id_list.length];

		for (int i = 0; i < id_list.length; i++) {
			Set<String> reportSet = reportsByUser.get(id_list[i]);
			if (reportSet != null) {
				answer[i] = (int)reportSet.stream().filter(blockedUser::contains).count();
			}
		}

		return answer;
	}
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
from typing import List

def solution(id_list: List[str], report: List[str], k: int):
    banned_users: set[str] = set()
    user_reported: dict[str, set[str]] = {
        id: set()
        for id in id_list
    }
    reported_counts: dict[str, int] = {
        id: 0
        for id in id_list
    }
    
    for entry in report:
        report_detail = entry.split(" ")
        reporter = report_detail[0]
        reported = report_detail[1]
        
        if not reported in user_reported[reporter]:
            user_reported[reporter].add(reported)
            reported_counts[reported] += 1
            
            if reported_counts[reported] == k:
                banned_users.add(reported)

    answer = []
                
    for id in id_list:
        total_mails = 0
        
        for banned_user in banned_users:
            if banned_user in user_reported[id]:
                total_mails += 1
                
        answer.append(total_mails)
        
    return answer